gnd/src/main/java/com/google/android/gnd/ui/recorddetails/RecordDetailsViewModel.java

```diff
@@ -17,32 +17,124 @@
+  public final LiveData<String> toolbarSubtitle;
+  public final LiveData<String> formNameView;
 
   @Inject
   RecordDetailsViewModel(DataRepository dataRepository) {
     this.dataRepository = dataRepository;
-    this.record = new MutableLiveData<>();
+
+    this.recordDetailsRequest = BehaviorSubject.create();
+
+    this.recordStream =
+        recordDetailsRequest
+          .switchMap(a ->
+              this.dataRepository.getRecordDetails(a.getProjectId(), a.getFeatureId(), a.getRecordId()))
+          .toFlowable(BackpressureStrategy.LATEST);
+
+    //TODO: Refactor to expose the fetched record directly.
+    this.records = LiveDataReactiveStreams.fromPublisher(
+        recordStream
+    );
+
+    this.progressBarVisibility = LiveDataReactiveStreams.fromPublisher(
+        recordStream.map(this::toProgressBarVisibility)
+        .onErrorReturn(x -> View.GONE)
+    );
+
+    this.toolbarTitle = LiveDataReactiveStreams.fromPublisher(
+        recordStream.map(this::toToolbarTitle)
+        .onErrorReturn(x -> "")
+    );
+
+    this.toolbarSubtitle = LiveDataReactiveStreams.fromPublisher(
+        recordStream.map(this::toToolbarSubtitle)
+        .onErrorReturn(x -> "")
+    );
+
+    this.formNameView = LiveDataReactiveStreams.fromPublisher(
+        recordStream.map(this::toFormNameView)
+        .onErrorReturn(x -> "")
+    );
+
+    recordDetailsRequest.subscribe(args -> {
+      Log.d(TAG, "" + args);
+    });
+  }
+
+  public void getRecordDetailsRequest(RecordDetailsFragmentArgs args) {
+    this.recordDetailsRequest.onNext(args);
+  }
+
+  private Integer toProgressBarVisibility(Resource<Record> record) {
+    switch (record.operationState().get()) {
+      case LOADING:
+        return View.VISIBLE;
+      default:
+        return View.GONE;
+    }
+  }
+
+  private String toToolbarTitle(Resource<Record> record) {
+    switch (record.operationState().get()) {
+      case LOADED:
+        return record.data()
+          .map(r -> r.getFeature().getTitle())
+          .orElse("");
+      default:
+        return "";
+    }
   }
 
-  public LiveData<Resource<Record>> getRecord() {
-    return record;
+  private String toToolbarSubtitle(Resource<Record> record) {
+    switch (record.operationState().get()) {
+      case LOADED:
+        return
+          record.data()
+            .map(r -> r.getFeature().getSubtitle())
+            .orElse("");
+      default:
+        return "";
+    }
   }
 
-  public void loadRecordDetails(String projectId, String featureId, String recordId) {
-    disposeOnClear(
-        dataRepository
-            .getRecordDetails(projectId, featureId, recordId)
-            .subscribe(record::setValue));
+  private String toFormNameView(Resource<Record> record) {
+    switch (record.operationState().get()) {
+      case LOADED:
+        return
+          record.data()
+            .map(r -> r.getForm().getTitle())
+            .orElse("");
+      default:
+        return "";
+    }
```

Comment:
Resource has a method data() which returns Optional.empty() if no data is loaded, allowing you to simplify the above "to" methods. For example, for toFormNameView:
`record.data().map(Record::getForm).map(Record::getTitle).orElse("");`

toProgressBarVisibility could be simplified as:
`record.data().isPresent() ? View.VISIBLE : View.GONE`

Reply:
Sweet! Simplification complete :)